### DISCUSSION TITLE
podman on Mac host, third party repository scripts using :z/:Z

### DISCUSSION THREAD

**Discussion by ncdc** [containers/podman#14602] [General]
@rhatdan following up from #13727 here to avoid any more noise on that issue 😄 

I'm trying to run some `make` commands from the openshift/release repo, and they run pretty much everything in containers, and the bind mounts have `:z`. What do you think the best course of action should be?

1. Consumers (e.g. openshift/release) should look for a Mac host and skip `:Z/:z` in their scripts
2. podman should ignore `:Z/:z` if the host is Mac
3. Something else?

Thanks!

**Comment by rhatdan** [comment]
Well Option 2 is impossible, since SELinux would have no idea it is on a MAC. 
The issue is Podman gets ENOTSUP on Relabel. We could change Podman to drop the relabel on that ERRNO, but that would be a big change. And since the Relabel is for security purposes, not something I would be in favor of.

How would openshift/release know it is on a MAC?

**Reply by rhatdan** [reply]
OpenShift could try to change a label on a file system and see if it gets ENOSUP and then drop the :z.

**Reply by rhatdan** [reply]
File systems without seclabel, do not support labeling.

```
$ cat /proc/self/mounts | grep -v seclabel
proc /proc proc rw,nosuid,nodev,noexec,relatime 0 0
securityfs /sys/kernel/security securityfs rw,nosuid,nodev,noexec,relatime 0 0
efivarfs /sys/firmware/efi/efivars efivarfs rw,nosuid,nodev,noexec,relatime 0 0
bpf /sys/fs/bpf bpf rw,nosuid,nodev,noexec,relatime,mode=700 0 0
selinuxfs /sys/fs/selinux selinuxfs rw,nosuid,noexec,relatime 0 0
systemd-1 /proc/sys/fs/binfmt_misc autofs rw,relatime,fd=35,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=14739 0 0
fusectl /sys/fs/fuse/connections fusectl rw,nosuid,nodev,noexec,relatime 0 0
configfs /sys/kernel/config configfs rw,nosuid,nodev,noexec,relatime 0 0
/dev/nvme0n1p1 /boot/efi vfat rw,relatime,fmask=0077,dmask=0077,codepage=437,iocharset=ascii,shortname=winnt,errors=remount-ro 0 0
binfmt_misc /proc/sys/fs/binfmt_misc binfmt_misc rw,nosuid,nodev,noexec,relatime 0 0
sunrpc /var/lib/nfs/rpc_pipefs rpc_pipefs rw,relatime 0 0
gvfsd-fuse /run/user/3267/gvfs fuse.gvfsd-fuse rw,nosuid,nodev,relatime,user_id=3267,group_id=3267 0 0
portal /run/user/3267/doc fuse.portal rw,nosuid,nodev,relatime,user_id=3267,group_id=3267 0 0
```

**Reply by rhatdan** [reply]
As opposed to 
```
$ cat /proc/self/mounts | grep seclabel
sysfs /sys sysfs rw,seclabel,nosuid,nodev,noexec,relatime 0 0
devtmpfs /dev devtmpfs rw,seclabel,nosuid,size=4096k,nr_inodes=1048576,mode=755,inode64 0 0
tmpfs /dev/shm tmpfs rw,seclabel,nosuid,nodev,inode64 0 0
devpts /dev/pts devpts rw,seclabel,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000 0 0
tmpfs /run tmpfs rw,seclabel,nosuid,nodev,size=13120292k,nr_inodes=819200,mode=755,inode64 0 0
cgroup2 /sys/fs/cgroup cgroup2 rw,seclabel,nosuid,nodev,noexec,relatime,nsdelegate,memory_recursiveprot 0 0
pstore /sys/fs/pstore pstore rw,seclabel,nosuid,nodev,noexec,relatime 0 0
/dev/mapper/fedora_localhost--live-root / xfs rw,seclabel,relatime,attr2,inode64,logbufs=8,logbsize=32k,noquota 0 0
mqueue /dev/mqueue mqueue rw,seclabel,nosuid,nodev,noexec,relatime 0 0
hugetlbfs /dev/hugepages hugetlbfs rw,seclabel,relatime,pagesize=2M 0 0
debugfs /sys/kernel/debug debugfs rw,seclabel,nosuid,nodev,noexec,relatime 0 0
tracefs /sys/kernel/tracing tracefs rw,seclabel,nosuid,nodev,noexec,relatime 0 0
tmpfs /tmp tmpfs rw,seclabel,nosuid,nodev,nr_inodes=1048576,inode64 0 0
/dev/nvme0n1p2 /boot ext4 rw,seclabel,relatime 0 0
/dev/mapper/luks-40793f09-de3b-4f0a-bb39-a4e5d90dc4b1 /home xfs rw,seclabel,relatime,attr2,inode64,logbufs=8,logbsize=32k,noquota 0 0
/dev/mapper/luks-1d46fdff-f599-4589-80c6-ff54591a6a44 /var xfs rw,seclabel,relatime,attr2,inode64,logbufs=8,logbsize=32k,noquota 0 0
/dev/mapper/luks-f9fdc088-ae95-49d3-89b2-d90861eac03b /var/lib/container xfs rw,seclabel,relatime,attr2,inode64,logbufs=8,logbsize=32k,noquota 0 0
tmpfs /run/user/3267 tmpfs rw,seclabel,nosuid,nodev,relatime,size=6560144k,nr_inodes=1640036,mode=700,uid=3267,gid=3267,inode64 0 0
```

**Reply by ncdc** [reply]
> Well Option 2 is impossible, since SELinux would have no idea it is on a MAC.

Couldn't the client portion of podman know that it's running on a Mac and silently drop the :z/:Z bit?

> How would openshift/release know it is on a MAC?

Note, this isn't specific to openshift/release. That's just where I happened to hit the problem. But in general, usually you check something like `uname -s` and look for `Darwin`.

**Reply by rhatdan** [reply]
What if you have a file system that supports labeling. you would want the :Z to work.  This will also work eventually once we move to virtiosfsd.

**Reply by ncdc** [reply]
But would that work for a Mac host?

**Reply by rhatdan** [reply]
virtiofsd on a mac host with relabling in the VM would work.

**Reply by ncdc** [reply]
So on a Mac, `podman run -v /a:/b:z` would work in that case? I'm thinking about the user experience, where someone is doing something similar to me - I grabbed a 3rd party repo, tried to run one of its scripts, and it failed. I could ask the repo maintainers to fix it (or submit a PR), but then I might run into the same issue in some other 3rd party repo potentially. Would it be worthwhile to build support into the podman client to ignore :z/:Z when the host is a Mac?

**Reply by rhatdan** [reply]
I don't believe so.  You told Podman to relabel content in these directories, it got an error saying "It is not supported" Podman on a MAC would not know if this is expected or not.

**Reply by ncdc** [reply]
So eventually a bind mounted volume coming from a Mac host with SELinux relabeling is something that will be supported (via virtiofsd)?

**Reply by rhatdan** [reply]
Yes.
